home-assistant/core
# Migrate decora_wifi (Leviton ) to UI ConfigFlow

## Breaking change

Migrate `decora_wifi` from platform config yaml to ConfigFlow via UI.

## Proposed change

[Leviton Decora Wifi](https://www.home-assistant.io/integrations/decora_wifi/) currently requires manual configuration, and only "entities" are added. This PR:
* Add Config Flow for Decora Wifi Integration, including validation of credentials
* Adds `DeviceInfo`, including fields previously not read by integration
* Wrapped Client calls into thin async client wrapper for clarify

If this change is green light, I will prepare a PR for homeassistant.io docs as well. 
We may be able to re-use [this PR](https://github.com/home-assistant/home-assistant.io/pull/22005))

#### Previous PR
Realized after creating this PR that similar PR was created last year, although it was never completed. I think they are similar, however decided to keep mine since the other was 1.5 years old and used some custom pattern vs established ones (e.g. used "trigger_import" vs "step_import" for importing yaml)
https://github.com/home-assistant/core/pull/68093/files

#### Additional Notes
1. MyLeviton uses "email" for authentication, but to improve backwards compatibility, keeping config key  as "USERNAME".

2. I am wondering if the correct entity for this integration should have been `Switch` and not `Light` since it control "Switches". Note even the api call is "get_iot_switches". It's likely they are controlling "lights", but also entirely possible they are controlling "outlets" or other devices. If we were to use the type switch instead, users could map switch to lights using "Switch As X" [integration](https://www.home-assistant.io/integrations/switch_as_x/). 
Obviously this change would be out of scope for this PR but curious if reviewers have any thoughts on this, and if it makes sense to consider or best way to handle type changes like this.

#### Code Owner
According to [adr-0008](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md), this might be considered a new feature or substantial refactor and therefore call for being added to `CODEOWNERS`.
Please let me know if this is not the case


## Type of change



## Additional information


- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: We may be able to re-use this one https://github.com/home-assistant/home-assistant.io/pull/22005/files


## Checklist


- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)

If user exposed functionality or configuration variables are added/changed:

If the code communicates with devices, web services, or third-party tools:

      Updated and included derived files by running: `python3 -m script.hassfest`.
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.



To help with the load of incoming pull requests:


[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure


[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr


### Tests

1. Setup account, ensured device was loaded and successfully controlled. Deleted and re-added.
2. Ran with credentials in `configuration.yaml`, ensured config was migrated into new integration
3. Ensure deprecation message was shown for the case above
4. Attempted to add account twice with same email, got expected "already registered error" 

<img width="1165" alt="image" src="https://github.com/home-assistant/core/assets/9513968/8a3003af-502e-46f1-9f46-1fd58ed8f12f">

<img width="1005" alt="image" src="https://github.com/home-assistant/core/assets/9513968/5de9595e-93d0-4377-9261-923965c650b0">
